hack/generator/pkg/astmodel/function.go

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package astmodel
+
+import (
+	"go/ast"
+)
+
+// Function represents something that is an (unnamed) Go function
+type Function interface {
+	HasImports
+	ReferenceChecker
+
+	// Renders the current instance as a Go abstract syntax tree
+	AsFunc(receiver *StructReference) *ast.FuncDecl
+}
+
+// FunctionDefinition is a specific function (or method, if receiver is not nil)
+type FunctionDefinition struct {
+	Function
+	DefinitionName
+
+	Receiver *StructReference
+}
+
+// Reference gets a reference to the function
+func (f *FunctionDefinition) Reference() *DefinitionName {
+	return &f.DefinitionName
+}
+
+// Tidy cleans up the definition prior to code generation
+func (f *FunctionDefinition) Tidy() {
+	return
+}
+
+// FileNameHint returns what a file that contains this definition (if any) should be called
+// this is not always used as we might combine multiple definitions into one file
+func (f *FunctionDefinition) FileNameHint() string {
+	return f.Name()
+}
+
+// AsDeclarations generates the Go code representing this definition
+func (f *FunctionDefinition) AsDeclarations() []ast.Decl {
+	result := f.Function.AsFunc(f.Receiver) // Get the func
+	result.Name = ast.NewIdent(f.Name())    // Give it a name
+
+	return []ast.Decl{result}
+}
+
+// References indicates whether this function includes any direct references to the given Type
+func (f *FunctionDefinition) References(t Type) bool {
+	// This is only used at the moment to group definitions into files so just returning false
+	return false
```

Comment:
This implementation makes sense for `IsReferencedBy()` - but as we discussed, adding `IsMemberOf()` would help allocate functions to files.

Reply:
Alternatively, using `References()` in both directions when allocating definitions to files may be better than adding another method.